src/audio/google/google_ctc_audio_processing.c

```diff
@@ -0,0 +1,574 @@
+			 struct output_stream_buffer *output_buffers,
+			 uint32_t frames);
+
+struct ctc_func_map {
+	enum sof_ipc_frame fmt; /**< source frame format */
+	ctc_func func; /**< processing function */
+};
+
+struct google_ctc_audio_processing_comp_data {
+	float *input;
+	float *output;
+	uint32_t input_ptr;
+	uint32_t output_ptr;
+	uint32_t num_frames;
+	uint32_t num_channels;
+	GoogleCtcAudioProcessingState *state;
+	struct comp_data_blob_handler *tuning_handler;
+	bool reconfigure;
+	ctc_func ctc_func;
+};
+
+static inline int16_t convert_float_to_int16(float data)
+{
+#if XCHAL_HAVE_HIFI3
+	const xtfloat ratio = 2 << 15;
+	xtfloat x0 = data;
+	xtfloat x1;
+	int16_t x;
+
+	x1 = XT_MUL_S(x0, ratio);
+	x = XT_TRUNC_S(x1, 0);
+
+	return x;
+#else /* XCHAL_HAVE_HIFI3 */
+	return Q_CONVERT_FLOAT(data, 15);
+#endif /* XCHAL_HAVE_HIFI3 */
+}
+
+static inline float convert_int16_to_float(int16_t data)
+{
+#if XCHAL_HAVE_HIFI3
+	const xtfloat ratio = 2 << 15;
+	xtfloat x0 = data;
+	float x;
+
+	x = XT_DIV_S(x0, ratio);
+
+	return x;
+#else /* XCHAL_HAVE_HIFI3 */
+	return Q_CONVERT_QTOF(data, 15);
+#endif /* XCHAL_HAVE_HIFI3 */
+}
+
+static inline int32_t convert_float_to_int32(float data)
+{
+#if XCHAL_HAVE_HIFI3
+	const xtfloat ratio = 2 << 15;
+	xtfloat x0 = data;
+	xtfloat x1;
+	int32_t x;
+
+	x1 = XT_MUL_S(x0, ratio);
+	x = XT_TRUNC_S(x1, 0);
+
+	return x;
+#else /* XCHAL_HAVE_HIFI3 */
+	return Q_CONVERT_FLOAT(data, 15);
+#endif /* XCHAL_HAVE_HIFI3 */
+}
+
+static inline float convert_int32_to_float(int32_t data)
+{
+#if XCHAL_HAVE_HIFI3
+	const xtfloat ratio = 2 << 15;
+	xtfloat x0 = data;
+	float x;
+
+	x = XT_DIV_S(x0, ratio);
+
+	return x;
+#else /* XCHAL_HAVE_HIFI3 */
+	return Q_CONVERT_QTOF(data, 15);
+#endif /* XCHAL_HAVE_HIFI3 */
+}
+
+static void ctc_s16_default(struct google_ctc_audio_processing_comp_data *cd,
+			    const struct audio_stream *source,
+			    struct audio_stream *sink,
+			    struct input_stream_buffer *input_buffers,
+			    struct output_stream_buffer *output_buffers,
+			    uint32_t frames)
+{
+	int n_ch = audio_stream_get_channels(source);
+	int samples = frames * n_ch;
+
+	int32_t *src = audio_stream_get_rptr(source);
+	int32_t *dest = audio_stream_get_wptr(sink);
+
+	int samples_to_process = MIN(samples, audio_stream_samples_without_wrap_s16(source, src));
+	int samples_to_consume = 0;
+
+	for (int i = 0; i < samples_to_process; ++i)
+		cd->input[cd->input_ptr++] = convert_int16_to_float(src[i]);
+	src = audio_stream_wrap(source, src + samples_to_process);
+	input_buffers->consumed += audio_stream_frame_bytes(source) * samples_to_process / n_ch;
+
+	while (cd->input_ptr >= cd->num_frames * n_ch) {
+		GoogleCtcAudioProcessingProcess(cd->state, cd->input, cd->output + cd->output_ptr,
+						cd->num_frames, n_ch);
+		cd->output_ptr += cd->num_frames * n_ch;
+
+		cd->input_ptr -= cd->num_frames * n_ch;
+		for (int i = 0; i < cd->input_ptr; ++i)
+			cd->input[i] = cd->input[i + cd->num_frames * n_ch];
```

Comment:
this seems rather inefficient - moving samples to the beginning of both input and output buffers (below). Maybe you could make this a circular buffer?

Reply:
Redesign the buffer to contain exactly `num_frames * n_ch` and consume them as soon as there are enough samples to pass to the CTC function.
In this way, the samples don't need to be moved.
Thanks for pointing this out!